clients/lsp-rust.el

```diff
@@ -928,5 +937,63 @@ them with `crate` or the crate name they refer to."
   (interactive)
   (lsp-find-locations "experimental/parentModule" nil :display-action display-action))
 
+(defun lsp-rust-analyzer-open-cargo-toml (&optional new-window)
+  "Open the closest Cargo.toml from the current file.
+
+   Rust-Analyzer LSP protocol documented here and added in November 2020
+   https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#open-cargotoml
+
+   If NEW-WINDOW (interactively the prefix argument) is non-nil,
+   open in a new window."
+  (interactive "P")
+  (-if-let (workspace (lsp-find-workspace 'rust-analyzer))
+      (-if-let* ((response (with-lsp-workspace workspace
+                             (lsp-send-request (lsp-make-request
+                                                "experimental/openCargoToml"
+                                                (lsp-make-rust-analyzer-open-cargo-toml-params
+                                                 :text-document (lsp--text-document-identifier))))))
+                 ((&Location :uri :range) response))
+          (funcall (if new-window #'find-file-other-window #'find-file)
+                   (lsp--uri-to-path uri))
+        (lsp--warn "Couldn't find a Cargo.toml file or your version of rust-analyzer doesn't support this extension"))
+    (lsp--error "OpenCargoToml is an extension available only with rust-analyzer")))
+
+
+(defun lsp-rust-analyzer--related-tests ()
+  "Get runnable test items related to the current TextDocumentPosition.
+Calls a rust-analyzer LSP extension endpoint that returns a wrapper over Runnable[]"
+  (lsp-send-request (lsp-make-request
+                     "rust-analyzer/relatedTests"
+                     (lsp--text-document-position-params))))
+
+(defun lsp-rust-analyzer--select-related-test ()
+  "Call the endpoint and ask for user selection.
+
+Cannot reuse `lsp-rust-analyzer--select-runnable' because the runnables endpoint
+responds with Runnable[], while relatedTests responds with TestInfo[], which is a wrapper
+over runnable. Also, this method doesn't set the `lsp-rust-analyzer--last-runnable' variable"
+  (-if-let* ((resp (lsp-rust-analyzer--related-tests))
+             (runnables (seq-map
+                         (-lambda ((&rust-analyzer:RelatedTests :runnable)) runnable)
+                         resp)))
+      ;; TODO Is it possible to bind
+      ;; C-SPC preview similar to what ivy-configured doom has with
+      ;; ivy-call-and-recenter
```

Comment:
@kiennq I am a doom emacs user with the ivy completion layer and I really like the C-SPC preview feature.

Considering each runnable has a location field, we can use it to jump to the start of the related test function definition before running it. 

Are there any generic, non-ivy APIs in emacs that we can use to implement that?

If not, can we advise the ivy module in doom to do that? 

I will delete the comment once we know how to proceed.

Reply:
moving to lsp-ivy